### ISSUE TITLE
Make creation of java:comp/env threadsafe

### ISSUE THREAD

**Issue by janbartel** [eclipse/jetty.project#3729]
Based on a report to jetty-dev:  https://www.eclipse.org/lists/jetty-dev/msg03326.html

The jndi context methods are not, according to the api, threadsafe. However, we should make the creation of the java:comp/env context threadsafe to support the deployment of multiple servers/webapps via different threads because  apps will not be able to do their own synchronization at the point at which this special context is created.

**Comment by mattsheppard** [comment]
Just popping some details that were on the mailing list here for the record:

Jan reported that the reproduction in https://github.com/mattsheppard/jetty-NameNotFoundException-remaining-name-env-reproduction was throwing an `java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached` rather than the expected `NameNotFoundException; remaining name 'env'`.

I was able to reproduce that OOM by dropping the number of threads to 1 and letting it run for a while - Embarrassingly I wasn't shutting down the ExecutorService I was using to start the two servers in parallel. I've just pushed a change to github that addresses that and I now seem to be able to run it with one server-starting-thread indefinitely and still see the `NameNotFoundException` when starting the servers with multiple threads, so please give it another try if you can.

**Comment by joakime** [comment]
Incidentally, and this is not related to your issue, but you don't have to start 2 `Server` instances to accomplish your end goals.
Just have named ServerConnectors and use a ContextHandler (or ServletContextHandler, or WebAppContext) with virtual hosts configured using `@<name>` syntax to specify which connector goes to which context.

1 Server,  multiple connectors, each on their own webapp.

**Comment by janbartel** [comment]
Matt, we've got some changes that would be great if you could test, as I still haven't been able to reproduce the threading issue running locally.  Could you checkout and build the branch jetty-9.4.x-3729-concurrent-naming-context, and then try it and see if that fixes your issue?

**Comment by mattsheppard** [comment]
I've run it a few times and I haven't seen that same `NameNotFoundException` so that's certainly positive - Thanks for looking into it!

Leaving that reproduction case running for a while I have still seen it apparently fail to deploy one of the webapps, but much less frequently than before and without any stack trace in jetty's log. When I get some time I'll try to dig further into what's still failing and turn it into a new issue if I can get any useful leads on the cause.

Thanks again,
Matt

**Comment by janbartel** [comment]
OK, I will close this issue. If you can track down any further problem with multithreaded deployment, please go ahead and open a new issue.
